src/ast/irbuilderbpf.cpp

```diff
@@ -830,6 +830,103 @@ CallInst *IRBuilderBPF::CreateGetNs(bool boot_time, const location &loc)
   return CreateHelperCall(fn, gettime_func_type, {}, "get_ns", &loc);
 }
 
+Value *IRBuilderBPF::CreateIntegerArrayCmp(Value *ctx,
+                                           Value *val1,
+                                           Value *val2,
+                                           const SizedType &array_type,
+                                           const bool inverse,
+                                           const location &loc)
+{
+  /*
+   // This function compares each character of the two arrays.
+   // It returns true if all are equal and false if any are different
+   // cmp([]char val1, []char val2)
+   {
+      for (size_t i = 0; i < n; i++)
+      {
+        if (val1[i] != val2[i])
+        {
+          return false;
+        }
+      }
+      return true;
+   }
+*/
+  auto elem_type = *array_type.GetElementTy();
+  const uint64_t n = array_type.GetSize();
+
+  Value *ll, *rr, *cmp;
+  AllocaInst *l = CreateAllocaBPF(elem_type, "ll");
+  AllocaInst *r = CreateAllocaBPF(elem_type, "rr");
+  Function *parent = GetInsertBlock()->getParent();
+  AllocaInst *store = CreateAllocaBPF(getInt1Ty(), "arraycmp.result");
+  BasicBlock *arr_ne = BasicBlock::Create(module_.getContext(),
+                                          "arraycmp.false",
+                                          parent);
+  BasicBlock *done = BasicBlock::Create(module_.getContext(),
+                                        "arraycmp.done",
+                                        parent);
+
+  CreateStore(getInt1(!inverse), store);
+  for (size_t i = 0; i < n; i++)
+  {
+    BasicBlock *arr_eq = BasicBlock::Create(module_.getContext(),
+                                            "arraycmp.loop",
+                                            parent);
+    if (onStack(array_type))
```

Comment:
This doesn't seem correct. One of the arrays may be on stack while the other one is not.

I'm also wondering if it'd make sense to make use of the `readDatastructElemFromStack` and `probereadDatastructElem` functions.

Reply:
> if it'd make sense to make use of the readDatastructElemFromStack and probereadDatastructElem functions.

This is doable, then we should be moving `readDatastructElemFromStack`  and `probereadDatastructElem` to `irbuilderbpf.cpp`. But the IRBuilderBPF doesn't have the `current_attach_point` variable, so I guess I need to pass this as a parameter. Is that alright?

Reply:
Yeah, that would be necessary and also, `expr_deleter_` is missing. So I guess that it's best to just not use those functions (they anyway do more functionality than what is necessary here).